### ISSUE TITLE
Feature Request: Alarms

### ISSUE THREAD

**Issue by Wolphin5** [Rem0o/FanControl.Releases#807]
I would love to be able to enable an alarm for if my water pump stops responding.  Although, it could (and should...) also be made for if a temp goes too high/low, water flow low/high (if system has a flow sensor), or for any other usable sensor high, low, out of range (high and low) with maybe a delay of seconds before triggering (for if there was just a sensor/software glitch).

Options for actions:

- Override fan controls to a preset level
- Audio warning
- Some sort of visual warning
- Shutting down the computer

Thinking further, if the individual alarms, just trigger, and then have alarm actions separate, so can also create "Mix" alarms with AND (all in alarm), OR (any in alarm), XOR (one but not both in alarm), or threshold (i.e. 4 alarms in, and need 3 in alarm before it triggers).  

Could also have the software highlight the alarms when the active, so can see at a glance what it is doing.

**Comment by Rem0o** [comment]
I don't hate the idea. Having an alarm as a card, like let say custom sensors, would kinda make sense.  1 type of card per alarm type, kinda like everything else in the software.  

It would be yet another thing to add from the Add button and to display. Might need to do something UX wise to make it all less cluttered, especially the add button.

**Comment by Wolphin5** [comment]
Nice, I hope to see it eventually :)
Less clutter... minimizing sections... hmmm... that gets the tech in me going on how that could work.  Even if the screen is fairly clean, compared to some programs, for what it shows.

- Controls to showing name, with % and RPM
- Speeds is already minimal (maybe hiding items if linked to control)
- Curves to name, and temp value and %

Then a minimizing to just showing the title for the sections?

**Comment by cowwoc** [comment]
Great ideas. The only thing I wanted to add is the option to reboot the computer on high temperature, not just shut it down.

I personally ran into an issue where CPU usage and temperature went through the roof for some inexplicable reason while I was away. In that case, I'd prefer the computer to reboot instead of shutting down because I've got some server-related stuff running on it. Also, I can't remote desktop into a computer that has been shut down.

If you want to get super fancy about it then I'd love to be able to configure "Reboot the computer when temperature reaches X, unless this happens Y times within Z minutes, in which case please shut down the computer."

**Comment by Wolphin5** [comment]
> If you want to get super fancy about it then I'd love to be able to configure "Reboot the computer when temperature surpasses X for a certain amount of time, unless this happens Y times within Z minutes, in which case please shut down the computer because it sounds like something is terribly wrong."

I like that idea!  able to have it reboot to rule out SW issue, before shutting down!

**Comment by Genda1ph** [comment]
Would love to see this implemented.
The barebones solution is to run a shell script and pass some parameters, this would allow usage of smth like [BurntToast](https://github.com/Windos/BurntToast) to show notiications.

**Comment by OomJan** [comment]
How about creating a new API enpoint for a custom action. Add a new plugin action element. On the UI assign a sensor, a plugin api action, a combo box for command, a threshold value and perhaps a parameter. I really need an alarm and I dont want to run another software to manage an alarm. My water loop popped once while in VR so I did not see it immediately.

I want to assign my flow meter and when it drops below 50rpm I want an alarm to sound and shut the system down.

```
namespace FanControl.Plugins
{
  /// <summary>Represents an action in the application.</summary>
  public interface IPluginAction
  {
    /// <summary>Unique id for the action.</summary>
    string Id { get; }

    /// <summary>Name that will be displayed to the user</summary>
    string Name { get; }

    /// <summary>List of actions provided</summary>
    List<String> Actions { get; }

    /// <summary>Index of the action that is selected. Default 0.</summary>
    Int ActionSelectedIndex { get; }

    /// <summary>Threshold of the sensor</summary>
    float? Value { get; }

    /// <summary>Custom parameter</summary>
    String Value { get; }
  }
}
```

**Comment by defaultsecurity** [comment]
@Rem0o 
Any plans to advance such features?

**Comment by elite4jonny** [comment]
@Rem0o I would also like to see this feature, it would be great to be notified of fan or pump failure before something catastrophic happens to a system

**Comment by OomJan** [comment]
@Rem0o If you trouble implementing this or just no time, it might be worth thinking of open sourcing the project and letting the community submit pull requests for new features like this one. This request has been open for 3 years and we are still waiting on an implementation.

**Comment by cheater** [comment]
I had a situation where this would have helped today. One fan had a cable stuck in the blades and it wasn't spinning, and I didn't know, it's in a place where you can't see it that well. I found out because FanControl had the icon that said "this control is not responding", but only because I was looking at it for some reason.

@Rem0o I suggest having a popup happen if a control stops responding. FanControl is brought to front, icon in the taskbar is highlighted, and a popup says the control stopped responding. This should also happen if the fault condition already exists at startup.

**Comment by fidoboy** [comment]
I really can't understand how it is possible that, more than 3 years after this petition was started, it has still not been taken into account or implemented in this otherwise wonderful program (in fact, the best in its category). What I know for sure is that this feature is very easy to add into the GUI and that it is a real must. Having a programmable acoustic procedure for temperature and any fan failure is almost mandatory. And it's as simple as adding the desired values for any sensor or any fan available.

The same than the preceding commentary I have recently suffered an issue with my GPU fan, but in my particular case it happened caused but some kind of driver problem (I solved it by simply reinstalling again the drivers and now the fan is working again). But the question is that FanControl didn't create any alert, warning, popup, shutdown of the system or notification. Total silence.

That's why I have decided to put here my two cents to may be help to enhance this useful and beautiful app. So PLEASE, consider it and add this feature into the main app or at least as a optional plugin. It could be REALLY useful to have an option to shutdown gracefully the system and/or create some acoustic alerts and/or any popup alert on screen with a warning. From my humble point of view it has no much sense to have such control with the fans controlling its speed and at the same time do not have any option available to setup a inmediate shutdown condition and some kind of acoustic/visual warnings.

**Comment by imniko** [comment]
I use HWinfo alerts, once my PC crashed due to GPU fan not spinning since then I have been using it, saved my ass so many times, but would be nice to have a feature like this in fan control of course.

**Comment by fidoboy** [comment]
> I use HWinfo alerts, once my PC crashed due to GPU fan not spinning since then I have been using it, saved my ass so many times, but would be nice to have a feature like this in fan control of course.

Yes. Because it's a waste to have two running apps when you can have only one for the same purppose. I believe that the alarm feature should be a must in this FanControl app. I can't understand why is not already added.

**Comment by cowwoc** [comment]
If you can program, contribute a pull-request. If you cannot, vote for this issue. Beyond that, complaining won't do much. The author is working on this software in his free time.

**Comment by Syl2010** [comment]
Isn't the source code closed? Doesn't seems possible to contribute.

**Comment by cowwoc** [comment]
@Syl2010 Good point. I didn't notice this :)

**Comment by tb0mb** [comment]
@Rem0o I love your software, I will donate $50 US if you add alarm function! If possible a shutdown command would be cool too, I don't really trust my computer alone 😜

**Comment by man55** [comment]
I agree with the request. I really like the program and have already donated.
But I would really like to see a safety feature like shutting down the PC when a trigger is triggered, such as when the GPU temperature exceeds the acceptable limit.

**Comment by tb0mb** [comment]
> [@Rem0o](https://github.com/Rem0o) I love your software, I will donate $50 US if you add alarm function! If possible a shutdown command would be cool too, I don't really trust my computer alone 😜

Mr. @Rem0o , I am prepared to send you $25 US today upon your acknowledgement of this request and $50 more after completion! I would love alarm function for over temp and critically, automatic shutdown. Thank you sir for your time 🙏

**Comment by a7dfj8aerj** [comment]
I have wrote on reddit and I will also add here 

I have looked at the case and realized the front fans werent spinning checked fan control and it had question mark and set to 100 percent but rpm was 0 and realized during harddrive replacement I accidently disconnected the Y splitter. 

Alarm feature will be very useful and definately add usability on the fan control app because these might happen or worse a pump might die and ability to set alarms is great feature

**Comment by tb0mb** [comment]
> > [@Rem0o](https://github.com/Rem0o) I love your software, I will donate $50 US if you add alarm function! If possible a shutdown command would be cool too, I don't really trust my computer alone 😜
> 
> Mr. [@Rem0o](https://github.com/Rem0o) , I am prepared to send you $25 US today upon your acknowledgement of this request and $50 more after completion! I would love alarm function for over temp and critically, automatic shutdown. Thank you sir for your time 🙏

friends, i want to share with you how i accomplished goal in hwinfo and share powershell code for maybe easier implementation for Mr. Rem0o. fwiw with both hwinfo and fancontrol auto starting, i had to have fan control delay start by 10 seconds for hwinfo sensors to all read correctly.

in hwinfo, run a program - powershell. arguments," powershell -windowstyle hidden -command {c:\users\tb0mb\documents\autoshutdown.ps1} "

sample of ps1,
C:\Windows\System32\shutdown.exe /s /t 11 /f /c " "; Add-Type -AssemblyName Microsoft.VisualBasic; $answer = [Microsoft.VisualBasic.Interaction]::MsgBox('Shutdown in 10 seconds. Cancel shutdown?','YesNo,SystemModal,Information', 'Auto Shutdown'); if ($answer -eq "Yes") { C:\Windows\System32\shutdown.exe -a}

I tried originally to do this all in the command string, but hit a character limit in hwinfo. hope it helps someone else! I'd still like to see fancontrol natively8 support alarming as i dont like having both apps open. i think it hurts performance double polling so many sensors and continue to see weird conflicting behavior between the 2 apps.
